src/main/scala/io/citrine/lolo/trees/splits/AnnealingSplitter.scala

```diff
@@ -0,0 +1,196 @@
+        cumSum = cumSum + score
+        if (draw < cumSum) {
+          return (new RealSplit(index, pivot), variance, totalScore)
+        }
+    }
+    throw new RuntimeException("Draw was beyond all the probabilities")
+  }
+
+  /**
+    * Get find the best categorical splitter.
+    *
+    * @param data        to split
+    * @param totalSum    Pre-computed data.map(d => data._2 * data._3).sum
+    * @param totalWeight Pre-computed data.map(d => d._3).sum
+    * @param index       of the feature to split on
+    * @return the best split of this feature
+    */
+  def getBestCategoricalSplit(
+                               data: Seq[(Vector[AnyVal], Double, Double)],
+                               totalSum: Double,
+                               totalWeight: Double,
+                               index: Int,
+                               minCount: Int,
+                               beta: Double
+                             ): (CategoricalSplit, Double, Double) = {
+    /* Extract the features at the index */
+    val thinData = data.map(dat => (dat._1(index).asInstanceOf[Char], dat._2, dat._3))
+
+    /* Group the data by categorical feature and compute the weighted sum and sum of the weights for each */
+    val groupedData = thinData.groupBy(_._1).mapValues(g => (g.map(v => v._2 * v._3).sum, g.map(_._3).sum, g.size))
+
+    /* Make sure there is more than one member for most of the classes */
+    val nonTrivial: Double = groupedData.filter(_._2._3 > 1).map(_._2._2).sum
+    if (nonTrivial / totalWeight < 0.5) {
+      return (new CategoricalSplit(index, BitSet()), Double.MaxValue, 0.0)
+    }
+
+    /* Compute the average label for each categorical value */
+    val categoryAverages: Map[Char, Double] = groupedData.mapValues(p => p._1 / p._2)
+
+    /* Create an orderd list of the categories by average label */
+    val orderedNames = categoryAverages.toSeq.sortBy(_._2).map(_._1)
+
+    /* Base cases for the iteration */
+    var leftSum = 0.0
+    var leftWeight = 0.0
+    var leftNum: Int = 0
+
+    /* Add the categories one at a time in order of their average label */
+    val possibleSplits: Seq[(Double, mutable.BitSet, Double)] = (0 until orderedNames.size - 1).map { j =>
+      val dat = groupedData(orderedNames(j))
+      leftSum = leftSum + dat._1
+      leftWeight = leftWeight + dat._2
+      leftNum = leftNum + dat._3
+
+      /* This is just relative, so we can subtract off the sum of the squares, data.map(Math.pow(_._2, 2)) */
+      val totalVariance = -leftSum * leftSum / leftWeight - Math.pow(totalSum - leftSum, 2) / (totalWeight - leftWeight)
+
+      val score = if (leftNum >= minCount && (thinData.size - leftNum) >= minCount) {
+        Math.exp(-totalVariance * beta)
+      } else {
+        0
+      }
+
+      val includeSet: mutable.BitSet = new mutable.BitSet() ++ orderedNames.slice(0, j + 1).map(_.toInt)
+
+      (score, includeSet, totalVariance)
+    }
+
+    val totalScore = possibleSplits.map(_._1).sum
```

Comment:
Is there any case where this totalScore would be zero?

Reply:
I changed the return value of the helper splitting methods to be options.  Now, in any case where a returned split would have zero probability, it actually returns None.  That way, we can "check for zero probability" by checking if the returned splits have any non-None (by flattening them and seeing if the result has any splits in it).